design/UsageValidationRules.md

```diff
@@ -0,0 +1,69 @@
+## Definition of "resources being used"
+
+Resources in WebGPU can be used in different ways that are declared at resource creation time and interact with various validation rules.
+
+Buffers have the following usages (and commands inducing that usage):
+ - `VERTEX` for the `buffers` in calls to `WebGPUCommandEncoder.setVertexBuffers`
+ - `INDEX` for `buffer` in calls to `WebGPUCommandEncoder.setIndexBuffer`
+ - `INDIRECT` for `indirectBuffer` in calls to `WebGPUCommandEncode.{draw|drawIndexed|dispatch}Indirect`
+ - `UBO`, `RO_STORAGE` and `STORAGE` for buffers referenced by bindgroups passed to `setBindGroup`, with the usage corresponding to the binding's type.
+ - `TRANSFER_SRC` for buffers used as the copy source of various commands.
+ - `TRANSFER_DST` for buffers used as the copy destination of various commands.
+ - (Maybe [RO_]STORAGE_TEXEL and UNIFORM_TEXEL?)
+
+Textures have the following usages:
+ - `OUTPUT_ATTACHMENT` for the subresources referenced by `WebGPURenderPassDescriptor`
+ - `SAMPLED`, `RO_STORAGE` and `STORAGE` for subresources corresponding to the image views referenced by bindgroups passed to `setBindGroup`, with the usage corresponding to the binding's type.
```

Comment:
What's the difference between `SAMPLED` and `RO_STORAGE`?

Reply:
In D3D12 speak `SAMPLED` is an SRV while `RO_STORAGE` is an UAV. Sampled always requires going through a sampler, while storage directly reads memory at a specific texel offset.

Reply:
I don't really see the reason for having all three usages. Is there some we can do with a read-only UAV but not with a SRV?

Reply:
I tend to agree - there doesn't appear to be a strong reason to have `SAMPLED` being a separate usage from `RO_STORAGE`

Reply:
Thinking about this more: `D3D12_RESOURCE_TRANSITION_BARRIER` doesn't allow `D3D12_RESOURCE_STATE_PIXEL_SHADER_RESOURCE` and `D3D12_RESOURCE_STATE_UNORDERED_ACCESS` at the same time so this probably means that we can only allow `RO_STORAGE` with `STORAGE` and nothing else. This would make `RO_STORAGE` only useful for type safety, which means we could remove it.

Reply:
I believe I might would make sense to rather have `RO_STORAGE` as additional descriptor type giving D3D12 the option to use SRVs and the Vulkan implementation to optimize UAV barriers a bit. But this kind of depends on the supported descriptor types and shading language.